storage/storage_image.go

```diff
@@ -717,16 +739,17 @@ func (s *storageImageDestination) Commit(ctx context.Context) error {
 	}
 	// Save the manifest.  Use storage.ImageDigestBigDataKey as the item's
 	// name, so that its digest can be used to locate the image in the Store.
-	if err := s.imageRef.transport.store.SetImageBigData(img.ID, storage.ImageDigestBigDataKey, s.manifest); err != nil {
+	if err := s.imageRef.transport.store.SetImageBigData(img.ID, storage.ImageDigestBigDataKey, destManifest); err != nil {
 		if _, err2 := s.imageRef.transport.store.DeleteImage(img.ID, true); err2 != nil {
 			logrus.Debugf("error deleting incomplete image %q: %v", img.ID, err2)
 		}
 		logrus.Debugf("error saving manifest for image %q: %v", img.ID, err)
 		return err
 	}
 	// Save the signatures, if we have any.
-	if len(s.signatures) > 0 {
-		if err := s.imageRef.transport.store.SetImageBigData(img.ID, "signatures", s.signatures); err != nil {
+	signatures := s.getSignatures()
```

Comment:
This seems overzealous and/or insufficient. What is the lock in `getSignatures` protecting against?

Just adding locks around individual field accesses does not automatically make code thread-safe:

If there could be any user possibly racing against `Commit` at this time and concurrently modifying `s.signatures`, that would mean `Commit` might be working with possibly obsolete / incorrect data (the other user might be adding more signatures after `Commit` has read them — whether or not the point-in-time _read_ of `s.signatures` was atomic or not. The fix in that case must be to wait for that other user to finish, not to add a lock.

(Similar concerns about a few other fields.)

As mentioned elsewhere, please document precisely what fields is each lock is intended to protect, and perhaps in which situations. I’d _maybe_ rather have 50 lines of correct comments than 5 lines of unnecessary/misleading code.

Reply:
I.e., I suspect that the `IsTreadSafe` property without further qualifications is not something that can _ever reasonably_ be answered ”yes”, it’s never “safe” to call `PutSignatures` concurrently with `Commit`, locking or not, or to call _anything_ concurrently with `Close`.

So, maybe restrict the “transport is tread-safe” property only to an explicit subset of the interface (`GetBlob`/`PutBlob`/`TryReusingBlob`), explicitly document that (both in the documentation of the `IsTreadSafe` method, which might benefit from a more restrictive name then, and in the documentation of the methods that may need to be safe when executed concurrently).

Reply:
It's only protecting against PutSignatues(). I will document things more precisely/carefully.
 
Note that by *thread-safe* I did not mean that anything can be executed in parallel. The order of execution must be controlled by the users. The biggest change is that PutBlob can be executed concurrently without screwing up state.

Reply:
I agree. Giving a false impression would be pretty bad. I'll update accordingly.